cert/cert.go

```diff
@@ -398,34 +434,34 @@ func (spec *Spec) Lifespan() time.Duration {
 	}
 
 	// This bit of code is necessary to confirm that the cert/key are older than the spec definition.
+	if spec.IsChangedOnDisk(spec.Key.Path) || spec.IsChangedOnDisk(spec.Cert.Path) {
+		// This is necessary to essentially force cfssl to regenerate since it's not spec aware.
+		log.Infof("refreshing due to spec %s having a newer mtime than key or cert", spec.Path)
+		spec.ResetLifespan()
+		return 0
+	}
+	return spec.tr.Lifespan()
+}
+
+func (spec *Spec) IsChangedOnDisk(path string) bool {
 	specStat, err := os.Stat(spec.Path)
 	if err != nil {
 		// The assertion here is that the spec actually
 		// exists. If it doesn't, something is wrong with the
 		// world.
-		panic("cert: certificate spec doesn't exist during RefreshKeys()")
+		panic("cert: isChangedOnDisk: failed to stat certificate spec")
 	}
-
-	isTooOld := func(path string) bool {
-		st, err := os.Stat(path)
-		if err != nil {
-			if os.IsNotExist(err) {
-				log.Errorf("while checking cert/key path %s, got path error %s", path, err)
-			}
-			return true
-		}
-		if specStat.ModTime().After(st.ModTime()) {
-			log.Infof("refreshing due to spec %s having a newer mtime then %s", spec.Path, path)
-			return true
+	st, err := os.Stat(path)
```

Comment:
Should this be a stat, or an Lstat?  Specifically, should certmgr tolerate resolving through a leaf node link?

My strong suspicion is 'no'- I've not traced the write pathway, but if that pathway doesn't atomically rename content into place, it would allow an attacker to drop a leaf symlink to redirect the cert to where they want.  That's extremely hard to utilize, but if the cert targetted is a cert used for discerning authentication, that could be a problem.

Reply:
I'm not sure I follow this -- is the concern that `path` is a symlink? And is the solution to return an error or something if it is?

Reply:
I'll follow up OOB for this.